### DISCUSSION TITLE
A simple code to generate the .seb file

### DISCUSSION THREAD

**Discussion by tamkeen-tms** [SafeExamBrowser/seb-server#48] [Q&A]
Hello.

We have a very short time to implement SEB into our existing LMS solution. We understand how the .SEB files are structured and formatted (XML in PList format, encrypted and then gziped). We also understand that the SEB server is the new way to issue/generate SEB config files via API. But we simply don't need all of that; **the plan is to simply generate (programmatically) a .SEB for each exam/student session, and ask the student to download it and open it via SEB. Now we cannot figure out how to do that!**

Started with a simple XML config like so
```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/
PropertyList-1.0.dtd">
<plist version="1.0">
    <dict>
        <key>allowDownUploads</key>
        <true/>
        <key>allowFlashFullscreen</key>
        <false/>
        <key>allowPreferencesWindow</key>
        <true/>
        <key>allowQuit</key>
        <true/>
        <key>allowSwitchToApplications</key>
        <true/>
        <key>blockPopUpWindows</key>
        <false/>
    </dict>
</plist>
```

Now, we need to encypt it, then gzip it, and simply push it to the trainee, right? But the SEB refuses to read it. 

Please, is there a simple way to do that, we want to push this as soon as possible. Thanks

**Comment by anhefti** [comment]
For simply create a SEB configuration file, please use the [SEB Configuration Tool](https://safeexambrowser.org/windows/win_usermanual_en.html#configuration)

**Reply by tamkeen-tms** [reply]
I hoped it was obvious that we want to do this programmatically. We need to create a new config for each exam/student session and ask the student to open it. Using PHP

**Reply by danschlet** [reply]
Maybe you can study the [Moodle SEB integration plugin](https://github.com/catalyst/moodle-quizaccess_seb), which creates configs using PHP. Note that this plugin doesn't encrypt the config files, which isn't necessary, as it calculates the [Config Key](https://safeexambrowser.org/developer/seb-config-key.html) for those generated settings and compares that key with the one sent by SEB clients in the HTTP request headers (and the next version will also support the SEB JavaScript API).

**Reply by tamkeen-tms** [reply]
@dansclet I did try to do that, but couldn't really find the function that generates the configs with in the code. Could you please refer to the code file? 

Thanks

**Reply by danschlet** [reply]
We didn't develop this Moodle plugin, ask in that repository.

**Reply by tamkeen-tms** [reply]
Okay, will give it another try then. Thanks :)

**Comment by anhefti** [comment]
For how to encrypt and package the plain plist file, please refer to the [SEB File Format Documentation](https://www.safeexambrowser.org/developer/seb-file-format.html)

Or if this is not working, please send us the error message and logs of SEB that refuses the configuration.

**Reply by tamkeen-tms** [reply]
I read this over and over, but it's beyond me, specially the whole encryption part. I want to use a password, but this too proved to be very complex to do (for me). My options right now are:

- A programmatic tool/lib to take in the settings and spit out the .seb file.
- A SEB server hosting ready to use right away.

Does either of these exist?

**Reply by danschlet** [reply]
We are aware that there are requests for a programmatic tool to create an encrypted .seb file from an unencrypted SEB XML config file. We unfortunately don't have resources to work on this. It could theoretically be created by some community developer using either the according parts of the .NET C# source code of SEB for Windows or the Obj-C code from the macOS/iOS version.

**Reply by tamkeen-tms** [reply]
I understand. I will do exactly that, once I figure this out, I will create a PHP package on push it to composer.

**Comment by anhefti** [comment]
> A programmatic tool/lib to take in the settings and spit out the .seb file.

This is the SEB Configuration Tool. There is no other official specific library in a specific language that does this.

> A SEB server hosting ready to use right away.

There is no public hosting or service for SEB Server yet. But you can [setup one](https://seb-server-setup.readthedocs.io/en/latest/overview.html) by yourself.

**Reply by tamkeen-tms** [reply]
I understand. Thanks for your help.
